### PR TITLE
Registration

### PR DESCRIPTION
**Pull request by mattydono**
Hi,

Sorry about the size of this PR but I had started working on parts before you showed me appropriate sizing. Going ahead I'll make sure to keep PR's relevant to the importance of the feature or fix.

- Added/Implemented Redux
- Added emotion

Thanks,

Matt

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -3,9 +3,17 @@
   "version": "0.1.0",
   "private": true,
   "dependencies": {
+    "@emotion/core": "^10.0.9",
+    "@emotion/styled": "^10.0.9",
+    "@types/react-redux": "^7.0.5",
```
Comment (Sharakai) [inline package.json]: `@types` packages should be in `devDependencies` - they are only used for working on, and building, the project, and not in the runtime code at all.

File: src/App.tsx [hunk 2]
```diff
@@ -1,13 +1,20 @@
 import * as React from 'react';
+import { Provider } from 'react-redux';
+import {Store} from 'redux';
+import { AppState, configureStore } from './redux';
+import {Registration} from './Registration/Registration'
 
 class App extends React.Component {
+
+  private store: Store<AppState> = configureStore();
```
Comment (Sharakai) [inline src/App.tsx:8]: We need to be more careful doing this, as it is transpiled to be in the `constructor`, which can be invoked multiple times in concurrent rendering (next version of React).
Doing it here also blocks rendering until the store is fully set up.

File: src/Registration/Registration.tsx [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+type DispatchProps = {
+    addUser: (user: User) => void
+}
+
+type State = {
+        email: string,
+        firstName: string,
+        surname: string,
+        type: string
```
Comment (Sharakai) [inline src/Registration/Registration.tsx]: This should likely be stricter typed, either as an enum or a string literal union.

File: src/Registration/Registration.tsx [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+                <button onClick={this.onPrint}>Print</button>
+            </div>
+        )
+    }
+
+    private onPrint = () => {
+        console.log(this.props.users)
+    }
+
+    private onTypeChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
```
Comment (Sharakai) [inline src/Registration/Registration.tsx]: You can nicer type this as...

```ts
private onTypeChange: React.ChangeEventHandler<HTMLSelectElement> = event => { ...
```

Same for the other handlers below.

File: src/Registration/redux/userReducer.tsx [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import {User} from '../User';
+import {ADD_USER, AddUserAction} from './userActions';
+
+export type UsersState = {
+    users: User[]
```
Comment (Sharakai) [inline src/Registration/redux/userReducer.tsx]: I would re-consider this being an array, and using a dictionary (object) instead, with an auto-generated uuid key - also include that ID as part of the User type.

File: tslint.json [hunk 6]
```diff
@@ -6,5 +6,10 @@
       "node_modules/**/*.ts",
       "coverage/lcov-report/*.js"
     ]
+  },
+  "rules": {
+    "interface-over-type-literal": false,
+    "no-console": false,
```
Comment (Sharakai) [inline tslint.json:15]: This shouldn't be here. `no-console` is a very good lint rule.
Same with the `class-name` rule.

These would have stopped it from compiling before, but I've changed the `defaultSeverity` to "warning" so it no longer will.

**Review by Sharakai — CHANGES_REQUESTED** [review]
Hopefully you can see the review now!

File: src/Registration/Registration.tsx (hunk 4 above)
Comment (mattydono) [reply src/Registration/Registration.tsx]: okay so this is essentially stating 

What we're listening for: the type of event = the event itself => {

?

File: src/Registration/redux/userReducer.tsx (hunk 5 above)
Comment (mattydono) [reply src/Registration/redux/userReducer.tsx]: would I need an ID for a user as their email addresses provided will be unique identifiers?

File: src/Registration/User.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+export type UserFirstName = string;
+export type UserSurname = string;
+export type UserEmail = string;
+export type UserType = string;
```
Comment (Sharakai) [inline src/Registration/User.ts]: You don't need to do all these types - It's very verbose, and doesn't add anything.
Just change the below `User` type to have them as strings...

```ts
export type User = {
  firstName: string;
  surname: string;
  email: string;
  type: UserType; (the enum)
}
```

File: src/Registration/redux/userReducer.tsx [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+export type UsersState = {
+  users: User[];
+};
+
+const initialState: UsersState = {
+  users: [],
+};
+
+export function usersReducer(
```
Comment (Sharakai) [inline src/Registration/redux/userReducer.tsx:12]: Why the decision not to use `redux-actions`' `handleActions`, seeing as you're already using `createAction` from the same library?

File: src/Registration/Registration.tsx (hunk 4 above)
Comment (Sharakai) [reply src/Registration/Registration.tsx]: It's retyping `onTypeChange` as a function which takes an event and returns nothing (`void`).
Previously, you only had the `event` typed, which doesn't strictly-type the return type of `void`

File: src/Registration/redux/userReducer.tsx (hunk 5 above)
Comment (Sharakai) [reply src/Registration/redux/userReducer.tsx]: In that case, you could use the `email` as the object key.
`Record<Email, User>;`

**Pull request closed without merging**